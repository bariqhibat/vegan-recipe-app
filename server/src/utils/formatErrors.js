import mon from "mongoose";
import { pick } from "lodash";

const formatErrors = (e) => {
  if (e instanceof mon.Error.ValidationError) {
    return Object.keys(e.errors).map((keyName) => {
      return pick(e.errors[keyName].properties, ["path", "message"]);
    });
  }

  return [{ path: "name", message: "something went wrong" }];
};

export default formatErrors;
