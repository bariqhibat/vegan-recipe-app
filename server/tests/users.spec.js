import {
  ME_QUERY,
  EMAIL_SIGNUP,
  EMAIL_LOGIN,
} from "../../web/src/graphql/Users";
import { mutate, query } from "./utils/providers";

const data = {
  user: {
    _id: null,
    firstName: "Bariq",
    lastName: "Nurlis",
    email: "bariqnurlis@bariq.com",
    password: "lklklklk",
  },
};

describe("Query", () => {
  describe("meQuery", () => {
    let userUUID = null;
    beforeEach(async () => {
      const res = await mutate(EMAIL_SIGNUP, {
        variables: {
          ...data.user,
        },
      });

      userUUID = res.data.createUser.user._id;
    });

    describe("Successful", () => {
      test("Through args", async () => {
        const res = await query(ME_QUERY, {
          variables: {
            userUUID,
          },
        });

        expect(res).toMatchObject({
          data: {
            meQuery: {
              ok: true,
              user: {
                _id: expect.any(String),
                firstName: expect.any(String),
                lastName: expect.any(String),
                email: expect.any(String),
              },
            },
          },
        });
      });
      // ! Unable since requires express middleware
      test("Through headers", async () => {
        //   const mock = await mutate(EMAIL_LOGIN, {
        //     variables: {
        //       ...data.user,
        //     },
        //   });
        //   const { token, refreshToken } = mock.data.login;
        //   setOptions({
        //     request: {
        //       headers: {
        //         "x-token": token,
        //         "x-refresh-token": refreshToken,
        //       },
        //     },
        //     response: {
        //       headers: {
        //         "x-token": token,
        //         "x-refresh-token": refreshToken,
        //       },
        //     },
        //   });
        //   const res = await query(ME_QUERY);
        //   expect(res).toMatchObject({
        //     data: {
        //       meQuery: {
        //         ok: true,
        //         user: {
        //           _id: expect.any(String),
        //           firstName: expect.any(String),
        //           lastName: expect.any(String),
        //           email: expect.any(String),
        //         },
        //       },
        //     },
        //   });
      });
    });
    describe("Unsuccessful", () => {
      test("Missing args", async () => {});
    });
  });
});
describe("Mutation", () => {
  describe("createUser", () => {
    describe("Successful", () => {
      test("Successful", async () => {
        const res = await mutate(EMAIL_SIGNUP, {
          variables: {
            ...data.user,
          },
        });

        expect(res).toMatchObject({
          data: {
            createUser: {
              ok: true,
              errors: null,
            },
          },
        });
      });
    });
    describe("Unsuccessful", () => {
      test("User already exists", async () => {
        await mutate(EMAIL_SIGNUP, {
          variables: {
            ...data.user,
          },
        });

        const res = await mutate(EMAIL_SIGNUP, {
          variables: {
            ...data.user,
          },
        });

        expect(res).toMatchObject({
          data: {
            createUser: {
              ok: false,
              user: null,
            },
          },
        });
      }),
        test("Missing args", async () => {
          await expect(async () => {
            await mutate(EMAIL_SIGNUP).catch((err) => {
              throw err;
            });
          }).rejects.toThrowErrorMatchingSnapshot();
        });
      describe("Wrong type of args", () => {
        it.each`
          email
          ${""}
          ${"bobby"}
          ${null}
          ${undefined}
        `("'$email' of email variant", async ({ email }) => {
          if (typeof email === "string") {
            const res = await mutate(EMAIL_SIGNUP, {
              variables: {
                ...data.user,
                email,
              },
            });

            expect(res).toMatchSnapshot();
            expect(res).toMatchObject({
              data: {
                createUser: {
                  ok: false,
                  user: null,
                },
              },
            });
          } else {
            await expect(async () => {
              await mutate(EMAIL_SIGNUP, {
                variables: {
                  ...data.user,
                  email,
                },
              }).catch((err) => {
                throw err;
              });
            }).rejects.toThrowErrorMatchingSnapshot();
          }
        });
      });
    });
  });
  describe("login", () => {
    beforeEach(async () => {
      await mutate(EMAIL_SIGNUP, {
        variables: {
          ...data.user,
        },
      });
    });

    describe("Successful", () => {
      test("Successful", async () => {
        const res = await mutate(EMAIL_LOGIN, {
          variables: {
            ...data.user,
          },
        });

        expect(res).toMatchObject({
          data: {
            login: {
              token: expect.any(String),
              refreshToken: expect.any(String),
              ok: true,
              user: {
                _id: expect.any(String),
              },
            },
          },
        });
      });
    });

    describe("Unsuccessful", () => {
      test("Missing args", async () => {
        await expect(async () => {
          await mutate(EMAIL_LOGIN).catch((err) => {
            throw err;
          });
        }).rejects.toThrowErrorMatchingSnapshot();
      });

      describe("Wrong type of args", () => {
        it.each`
          email
          ${""}
          ${"bobby"}
          ${null}
          ${undefined}
        `("'$email' of email variant", async ({ email }) => {
          if (typeof email === "string") {
            const res = await mutate(EMAIL_LOGIN, {
              variables: {
                ...data.user,
                email,
              },
            });

            expect(res).toMatchSnapshot();
            expect(res).toMatchObject({
              data: {
                login: {
                  ok: false,
                  user: null,
                  token: null,
                  refreshToken: null,
                },
              },
            });
          } else {
            await expect(async () => {
              await mutate(EMAIL_LOGIN, {
                variables: {
                  ...data.user,
                  email,
                },
              }).catch((err) => {
                throw err;
              });
            }).rejects.toThrowErrorMatchingSnapshot();
          }
        });
      });
    });
  });
});
