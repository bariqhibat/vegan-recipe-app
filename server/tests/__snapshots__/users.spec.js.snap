// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation createUser Unsuccessful Missing args 1`] = `
"{\\"errors\\":[{\\"message\\":\\"Variable \\\\\\"$firstName\\\\\\" of required type \\\\\\"String!\\\\\\" was not provided.\\",\\"locations\\":[{\\"line\\":1,\\"column\\":11}],\\"extensions\\":{\\"code\\":\\"INTERNAL_SERVER_ERROR\\"}},{\\"message\\":\\"Variable \\\\\\"$lastName\\\\\\" of required type \\\\\\"String!\\\\\\" was not provided.\\",\\"locations\\":[{\\"line\\":1,\\"column\\":32}],\\"extensions\\":{\\"code\\":\\"INTERNAL_SERVER_ERROR\\"}},{\\"message\\":\\"Variable \\\\\\"$email\\\\\\" of required type \\\\\\"String!\\\\\\" was not provided.\\",\\"locations\\":[{\\"line\\":1,\\"column\\":52}],\\"extensions\\":{\\"code\\":\\"INTERNAL_SERVER_ERROR\\"}},{\\"message\\":\\"Variable \\\\\\"$password\\\\\\" of required type \\\\\\"String!\\\\\\" was not provided.\\",\\"locations\\":[{\\"line\\":1,\\"column\\":69}],\\"extensions\\":{\\"code\\":\\"INTERNAL_SERVER_ERROR\\"}}]}
"
`;

exports[`Mutation createUser Unsuccessful Wrong type of args '' of email variant 1`] = `
Object {
  "data": Object {
    "createUser": Object {
      "errors": Array [
        Object {
          "message": "Email is required",
          "path": "email",
        },
      ],
      "ok": false,
      "user": null,
    },
  },
}
`;

exports[`Mutation createUser Unsuccessful Wrong type of args 'bobby' of email variant 1`] = `
Object {
  "data": Object {
    "createUser": Object {
      "errors": Array [
        Object {
          "message": "Invalid email format",
          "path": "email",
        },
      ],
      "ok": false,
      "user": null,
    },
  },
}
`;

exports[`Mutation createUser Unsuccessful Wrong type of args 'null' of email variant 1`] = `
"{\\"errors\\":[{\\"message\\":\\"Variable \\\\\\"$email\\\\\\" of non-null type \\\\\\"String!\\\\\\" must not be null.\\",\\"locations\\":[{\\"line\\":1,\\"column\\":52}],\\"extensions\\":{\\"code\\":\\"INTERNAL_SERVER_ERROR\\"}}]}
"
`;

exports[`Mutation createUser Unsuccessful Wrong type of args 'undefined' of email variant 1`] = `
"{\\"errors\\":[{\\"message\\":\\"Variable \\\\\\"$email\\\\\\" got invalid value undefined; Expected non-nullable type \\\\\\"String!\\\\\\" not to be null.\\",\\"locations\\":[{\\"line\\":1,\\"column\\":52}],\\"extensions\\":{\\"code\\":\\"INTERNAL_SERVER_ERROR\\"}}]}
"
`;

exports[`Mutation login Unsuccessful Missing args 1`] = `
"{\\"errors\\":[{\\"message\\":\\"Variable \\\\\\"$email\\\\\\" of required type \\\\\\"String!\\\\\\" was not provided.\\",\\"locations\\":[{\\"line\\":1,\\"column\\":11}],\\"extensions\\":{\\"code\\":\\"INTERNAL_SERVER_ERROR\\"}},{\\"message\\":\\"Variable \\\\\\"$password\\\\\\" of required type \\\\\\"String!\\\\\\" was not provided.\\",\\"locations\\":[{\\"line\\":1,\\"column\\":28}],\\"extensions\\":{\\"code\\":\\"INTERNAL_SERVER_ERROR\\"}}]}
"
`;

exports[`Mutation login Unsuccessful Wrong type of args '' of email variant 1`] = `
Object {
  "data": Object {
    "login": Object {
      "errors": Array [
        Object {
          "message": "Wrong email",
          "path": "email",
        },
      ],
      "ok": false,
      "refreshToken": null,
      "token": null,
      "user": null,
    },
  },
}
`;

exports[`Mutation login Unsuccessful Wrong type of args 'bobby' of email variant 1`] = `
Object {
  "data": Object {
    "login": Object {
      "errors": Array [
        Object {
          "message": "Wrong email",
          "path": "email",
        },
      ],
      "ok": false,
      "refreshToken": null,
      "token": null,
      "user": null,
    },
  },
}
`;

exports[`Mutation login Unsuccessful Wrong type of args 'null' of email variant 1`] = `
"{\\"errors\\":[{\\"message\\":\\"Variable \\\\\\"$email\\\\\\" of non-null type \\\\\\"String!\\\\\\" must not be null.\\",\\"locations\\":[{\\"line\\":1,\\"column\\":11}],\\"extensions\\":{\\"code\\":\\"INTERNAL_SERVER_ERROR\\"}}]}
"
`;

exports[`Mutation login Unsuccessful Wrong type of args 'undefined' of email variant 1`] = `
"{\\"errors\\":[{\\"message\\":\\"Variable \\\\\\"$email\\\\\\" got invalid value undefined; Expected non-nullable type \\\\\\"String!\\\\\\" not to be null.\\",\\"locations\\":[{\\"line\\":1,\\"column\\":11}],\\"extensions\\":{\\"code\\":\\"INTERNAL_SERVER_ERROR\\"}}]}
"
`;
